torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,202 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+from torch.utils import _pytree as pytree
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    resize_nodes_to_insert_at_end = []
+    node_to_usage_map = defaultdict(list)
+    for i, n in enumerate(node_list):
+        if n.op == "call_function":
+            if (
+                n.target is torch.ops.inductor.resize_storage_bytes_.default
+                and n.args[1] == 0
+            ):
+                resize_nodes_to_insert_at_end.append((n, i))
+            args_flat = pytree.arg_tree_leaves(*n.args, **n.kwargs)
+            for arg in args_flat:
+                node_to_usage_map[arg].append(i)
+    # Assumptions: the size-0 tensor X can then be used only in the following ops:
+    # 1. `aten.set_(X, ...)`
+    # 2. `aten.set_(..., X)`
+    # 3. `inductor.resize_storage_bytes_(X, 0)`
+    # 4. graph output
+    for resize_node, resize_node_idx in reversed(resize_nodes_to_insert_at_end):
+        usage_idx_list = node_to_usage_map[resize_node.args[0]]
+        pos_in_usage_idx_list = usage_idx_list.index(resize_node_idx)
+        for i in range(pos_in_usage_idx_list + 1, len(usage_idx_list)):
```

Comment:
just a nit (feel free to ignore), but I wonder if it'd be cleaner to loop over `resize_node.args[0].users` directly, and find the ones that show up after `resize_node` in the graph. I guess requires some boilerplate too, since you need a way to order the users

Reply:
yes great idea - I switched over to using `.users`